app/fvwm/modules/FvwmCpp/FvwmCpp.1
.\" $OpenBSD: FvwmCpp.1,v 1.1.1.1 2006/11/26 10:53:46 matthieu Exp $
.\" t
.\" @(#)FvwmCpp.1	1/28/94
.de EX		\"Begin example
.ne 5
.if n .sp 1
.if t .sp .5
.nf
.in +.5i
..
.de EE
.fi
.in -.5i
.if n .sp 1
.if t .sp .5
..
.TH FvwmCpp 1 12/12/94 2.0
.UC
.SH NAME
FvwmCpp \- the FVWM  Cpp pre-processor
.SH SYNOPSIS
FvwmCpp is spawned by fvwm, so no command line invocation will work.

.SH DESCRIPTION
When called, this module will attempt to have /usr/lib/cpp pre-process
the file specified in its invocation, and then have fvwm read the
resulting file.

.SH INVOCATION
FvwmCpp can be invoked by inserting the line 'FvwmCpp' in the .fvwmrc
file. It can also be called from a menu or mouse binding.  If the user
wants his entire .fvwmrc file pre-processed with FvwmCpp, then fvwm
should be invoked as:

.EX
fvwm2 -cmd "FvwmCpp .fvwmrc"
.EE

Some options can be specified on the command line:
.TP
-cppopt \fIoption\fP
Lets you pass an option to the cpp program.  Not really needed as any unknown
options will be passed on automatically.

.TP
-cppprog \fIname\fP
Instead of invoking "/usr/lib/cpp", fvwm will invoke \fIname\fP.

.TP
-outfile \fIfilename\fP
Instead of creating a random unique name for the temporary file for
the preprocessed rc file, this option will let you specify the name of
the temporary file it will create.

.IP -debug
Causes the temporary file create by Cpp to
be retained. This file is usually called "/tmp/fvwmrcXXXXXXXXXX"


.SH CONFIGURATION OPTIONS
FvwmCpp defines some values for use in the pre-processor file:

.IP TWM_TYPE
Always set to "fvwm".
.IP SERVERHOST
The name of the machine running the X Server.
.IP CLIENTHOST
The name of the machine running fvwm.
.IP HOSTNAME
The hostname of the machine running fvwm. Generally the same as CLIENTHOST.
.IP OSTYPE
The operating system for CLIENTHOST.
.IP USER
The name of the person running fvwm.
.IP VERSION
The X11 version.
.IP REVISION
The X11 revision number.
.IP VENDOR
The X server vendor.
.IP RELEASE
The X server release number.
.IP WIDTH
The screen width in pixels.
.IP HEIGHT
The screen height in pixels.
.IP X_RESOLUTION
Some distance/pixel measurement for the horizontal direction, I think.
.IP Y_RESOLUTION
Some distance/pixel measurement for the vertical direction, I think.
.IP PLANES
Number of color planes for the X server display
.IP BITS_PER_RGB
Number of bits in each rgb triplet.
.IP CLASS
The X11 visual class, ie PsuedoColor.
.IP COLOR
Yes or No, Yes if the display class is neither StaticGrey or GreyScale.
.IP FVWM_VERSION
The fvwm version number, ie 2.0
.IP OPTIONS
Some combination of SHAPE, XPM, NO_SAVEUNDERS, and Cpp, as defined in
configure.h at compile time.
.IP FVWM_MODULEDIR
The directory where fvwm looks for .fvwmrc and modules by default, as
determined at compile time.

.SH EXAMPLE PROLOG

.EX
#define TWM_TYPE fvwm
#define SERVERHOST spx20
#define CLIENTHOST grumpy
#define HOSTNAME grumpy
#define OSTYPE SunOS
#define USER nation
#define HOME /local/homes/dsp/nation
#define VERSION 11
#define REVISION 0
#define VENDOR HDS human designed systems, inc. (2.1.2-D)
#define RELEASE 4
#define WIDTH 1280
#define HEIGHT 1024
#define X_RESOLUTION 3938
#define Y_RESOLUTION 3938
#define PLANES 8
#define BITS_PER_RGB 8
#define CLASS PseudoColor
#define COLOR Yes
#define FVWM_VERSION 2.0 pl 1
#define OPTIONS SHAPE XPM Cpp
#define FVWM_MODULEDIR /local/homes/dsp/nation/modules

.EE

.SH BUGS
Module configurations do not become active until fvwm2 has restarted
if you use FvwmCpp on startup. FvwmCpp creates a temporary file
and passes this to fvwm2, so you would have to edit this file too.
There are some problems with comments in your .fvwmrc file.
The comment sign # is misinterpreted by the preprocessor.
This has usually no impact on functionality but generates
annoying warning messages.
The sequence /* is interpreted as the start of a C comment what
is probably not what you want in a filename. You might want to try
/?* (for filenames only) or /\\* or "/*" instead.
Macros are not replaced within single (') or double quotes ("). Use
backquotes (`) to circumvent this. Fvwm2 accepts backquotes for
quoting and at least FvwmButtons does too.
The preprocessor may place a space after a macro substitution, so
with
.EX
#define MYCOMMAND ls
"Exec "MYCOMMAND" -l"
.EE
you might get
.EX
"Exec "ls " -l" (two words)
.EE
and not
.EX
"Exec "ls" -l" (one word).
.EE

.SH AUTHOR
FvwmCpp is the result of a random bit mutation on a hard disk,
presumably a result of a  cosmic-ray or some such thing.


app/fvwm/modules/FvwmCpp/FvwmCpp.c
/* This module, and the entire FvwmM4 program, and the concept for
 * interfacing this module to the Window Manager, are all original work
 * by Robert Nation
 *
 * Copyright 1994, Robert Nation
 *  No guarantees or warantees or anything
 * are provided or implied in any way whatsoever. Use this program at your
 * own risk. Permission to use this program for any purpose is given,
 * as long as the copyright is kept intact. */

#define TRUE 1
#define FALSE  0

#include "config.h"

#include <stdio.h>
#include <signal.h>
#include <fcntl.h>
#include <string.h>
#include <sys/wait.h>
#include <sys/time.h>
#include <unistd.h>
#include <ctype.h>
#include <stdlib.h>
#include <pwd.h>
#include <sys/param.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netdb.h>

#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/Xproto.h>
#include <X11/Xatom.h>
#include <X11/Intrinsic.h>

#include "../../fvwm/module.h"

#include "FvwmCpp.h"
#include "../../libs/fvwmlib.h"
#include <X11/StringDefs.h>
#include <X11/Intrinsic.h>
#include <X11/Shell.h>
#define Resolution(pixels, mm)  ((((pixels) * 100000 / (mm)) + 50) / 100)

char *MyName;
int fd[2];

struct list *list_root = NULL;

int ScreenWidth, ScreenHeight;
int Mscreen;

long Vx, Vy;
static char *MkDef(char *name, char *def);
static char *MkNum(char *name,int def);
static char *cpp_defs(Display *display, const char *host, char *m4_options, char *config_file);
#define MAXHOSTNAME 255
#define EXTRA 20

char *cpp_prog = FVWM_CPP;          /* Name of the cpp program */

char cpp_options[BUFSIZ];
char cpp_outfile[BUFSIZ]="";

/***********************************************************************
 *
 *  Procedure:

 *	main - start of module
 *
 ***********************************************************************/
int main(int argc, char **argv)
{
  Display *dpy;			/* which display are we talking to */
  char *temp, *s;
  char *display_name = NULL;
  char *filename = NULL;
  char *tmp_file, read_string[80],delete_string[80];
  int i,cpp_debug = 0;

  strcpy(cpp_options,"");

  /* Record the program name for error messages */
  temp = argv[0];

  s=strrchr(argv[0], '/');
  if (s != NULL)
    temp = s + 1;

  MyName = safemalloc(strlen(temp)+2);
  strcpy(MyName,"*");
  strcat(MyName, temp);

  if(argc < 6)
    {
      fprintf(stderr,"%s Version %s should only be executed by fvwm!\n",MyName,
	      VERSION);
      exit(1);
    }

  /* Open the X display */
  if (!(dpy = XOpenDisplay(display_name)))
    {
      fprintf(stderr,"%s: can't open display %s", MyName,
	      XDisplayName(display_name));
      exit (1);
    }


  Mscreen= DefaultScreen(dpy);
  ScreenHeight = DisplayHeight(dpy,Mscreen);
  ScreenWidth = DisplayWidth(dpy,Mscreen);

  /* We should exit if our fvwm pipes die */
  signal (SIGPIPE, DeadPipe);

  fd[0] = atoi(argv[1]);
  fd[1] = atoi(argv[2]);

  for(i=6;i<argc;i++)
    {
      /* leaving this in just in case-- any option starting with '-'
         will get passed on to cpp anyway */
      if(strcasecmp(argv[i],"-cppopt") == 0)
	{
	  strlcat(cpp_options, argv[++i], sizeof(cpp_options));
	  strlcat(cpp_options, " ", sizeof(cpp_options));
	}
      else if (strcasecmp(argv[i], "-cppprog") == 0)
	{
	  cpp_prog = argv[++i];
	}
      else if(strcasecmp(argv[i], "-outfile") == 0)
      {
        strlcpy(cpp_outfile,argv[++i],sizeof(cpp_outfile));
      }
      else if(strcasecmp(argv[i], "-debug") == 0)
	{
	  cpp_debug = 1;
	}
      else if (strncasecmp(argv[i],"-",1) == 0)
      {
        /* pass on any other arguments starting with '-' to cpp */
        strlcat(cpp_options, argv[i],sizeof(cpp_options));
        strlcat(cpp_options, " ",sizeof(cpp_options));
      }
      else
	filename = argv[i];
    }

  /* If we don't have a cpp, stop right now.  */
  if (!cpp_prog || cpp_prog[0] == '\0') {
    fprintf(stderr, "%s: no C preprocessor program specified\n", MyName);
    exit(1);
  }

  for(i=0;i<strlen(filename);i++)
    if((filename[i] == '\n')||(filename[i] == '\r'))
      {
	filename[i] = 0;
      }

  if (!(dpy = XOpenDisplay(display_name)))
    {
      fprintf(stderr,"%s: can't open display %s",
	      MyName, XDisplayName(display_name));
      exit (1);
    }

  tmp_file = cpp_defs(dpy, display_name,cpp_options, filename);

  snprintf(read_string,sizeof(read_string),"read %s\n",tmp_file);
  SendInfo(fd,read_string,0);

  /* For a debugging version, we may wish to omit this part. */
  /* I'll let some cpp advocates clean this up */
  if(!cpp_debug)
    {
      snprintf(delete_string,sizeof(delete_string),"exec rm %s\n",tmp_file);
      SendInfo(fd,delete_string,0);
    }
  return 0;
}



static char *cpp_defs(Display *display, const char *host, char *cpp_options, char *config_file)
{
  Screen *screen;
  Visual *visual;
  char client[MAXHOSTNAME], server[MAXHOSTNAME], *colon;
  char ostype[BUFSIZ];
  char options[BUFSIZ];
  static char tmp_name[BUFSIZ];
  struct hostent *hostname;
  char *vc;			/* Visual Class */
  FILE *tmpf;
  struct passwd *pwent;
  int fd;
  /* Generate a temporary filename.  Honor the TMPDIR environment variable,
     if set. Hope nobody deletes this file! */

  if (strlen(cpp_outfile) == 0) {
    if ((vc=getenv("TMPDIR"))) {
      strlcpy(tmp_name, vc, sizeof(tmp_name));
    } else {
      strlcpy(tmp_name, "/tmp", sizeof(tmp_name));
    }
    strlcat(tmp_name, "/fvwmrcXXXXXXXXXX", sizeof(tmp_name));
    mktemp(tmp_name);
  } else {
    strlcpy(tmp_name,cpp_outfile, sizeof(tmp_name));
  }

  if (*tmp_name == '\0')
    {
      perror("mktemp failed in cpp_defs");
      exit(0377);
    }

  /*
  ** check to make sure it doesn't exist already, to prevent security hole
  */
  if ((fd = open(tmp_name, O_WRONLY|O_EXCL|O_CREAT, 0600)) < 0)
  {
    perror("exclusive open for output file failed in cpp_defs");
    exit(0377);
  }
  close(fd);

    /*
     * Create the appropriate command line to run cpp, and
     * open a pipe to the command.
     */

  strlcpy(options, cpp_prog, sizeof(options));
  strlcat(options, " ", sizeof(options));
  strlcat(options, cpp_options, sizeof(options));
  strlcat(options, " >", sizeof(options));
  strlcat(options, tmp_name, sizeof(options));

  tmpf = popen(options, "w");
  if (tmpf == NULL) {
    perror("Cannot open pipe to cpp");
    exit(0377);
  }

  gethostname(client,MAXHOSTNAME);

  getostype  (ostype, sizeof ostype);

  hostname = gethostbyname(client);
  strlcpy(server, XDisplayName(host), sizeof(server));
  colon = strchr(server, ':');
  if (colon != NULL) *colon = '\0';
  if ((server[0] == '\0') || (!strcmp(server, "unix")))
    strlcpy(server, client, sizeof(server));	/* must be connected to :0 or unix:0 */

  /* TWM_TYPE is fvwm, for completeness */

  fputs(MkDef("TWM_TYPE", "fvwm"), tmpf);

  /* The machine running the X server */
  fputs(MkDef("SERVERHOST", server), tmpf);
  /* The machine running the window manager process */
  fputs(MkDef("CLIENTHOST", client), tmpf);
  if (hostname)
    fputs(MkDef("HOSTNAME", (char *)hostname->h_name), tmpf);
  else
    fputs(MkDef("HOSTNAME", (char *)client), tmpf);

  fputs(MkDef("OSTYPE", ostype), tmpf);

  pwent=getpwuid(geteuid());
  fputs(MkDef("USER", pwent->pw_name), tmpf);

  fputs(MkDef("HOME", getenv("HOME")), tmpf);
  fputs(MkNum("VERSION", ProtocolVersion(display)), tmpf);
  fputs(MkNum("REVISION", ProtocolRevision(display)), tmpf);
  fputs(MkDef("VENDOR", ServerVendor(display)), tmpf);
  fputs(MkNum("RELEASE", VendorRelease(display)), tmpf);
  screen = ScreenOfDisplay(display, Mscreen);
  visual = DefaultVisualOfScreen(screen);
  fputs(MkNum("WIDTH", DisplayWidth(display,Mscreen)), tmpf);
  fputs(MkNum("HEIGHT", DisplayHeight(display,Mscreen)), tmpf);

  fputs(MkNum("X_RESOLUTION",Resolution(screen->width,screen->mwidth)),tmpf);
  fputs(MkNum("Y_RESOLUTION",Resolution(screen->height,screen->mheight)),tmpf);
  fputs(MkNum("PLANES",DisplayPlanes(display, Mscreen)), tmpf);

  fputs(MkNum("BITS_PER_RGB", visual->bits_per_rgb), tmpf);
  fputs(MkNum("SCREEN", Mscreen), tmpf);

  switch(visual->class)
    {
    case(StaticGray):
	  vc = "StaticGray";
	break;
	case(GrayScale):
	  vc = "GrayScale";
	break;
	case(StaticColor):
	  vc = "StaticColor";
	break;
	case(PseudoColor):
	  vc = "PseudoColor";
	break;
	case(TrueColor):
	  vc = "TrueColor";
	break;
	case(DirectColor):
	  vc = "DirectColor";
	break;
      default:
	vc = "NonStandard";
	break;
      }

    fputs(MkDef("CLASS", vc), tmpf);
    if (visual->class != StaticGray && visual->class != GrayScale)
      fputs(MkDef("COLOR", "Yes"), tmpf);
    else
      fputs(MkDef("COLOR", "No"), tmpf);
    fputs(MkDef("FVWM_VERSION", VERSION), tmpf);

    /* Add options together */
    *options = '\0';
#ifdef	SHAPE
    strcat(options, "SHAPE ");
#endif
#ifdef	XPM
    strcat(options, "XPM ");
#endif

    strcat(options, "Cpp ");

#ifdef	NO_SAVEUNDERS
    strcat(options, "NO_SAVEUNDERS ");
#endif

    fputs(MkDef("OPTIONS", options), tmpf);

    fputs(MkDef("FVWM_MODULEDIR", FVWM_MODULEDIR), tmpf);
    fputs(MkDef("FVWM_CONFIGDIR", FVWM_CONFIGDIR), tmpf);

    /*
     * At this point, we've sent the definitions to cpp.  Just include
     * the fvwmrc file now.
     */

    fprintf(tmpf, "#include \"%s\"\n", config_file);

    pclose(tmpf);
    return(tmp_name);
}





/***********************************************************************
 *
 *  Procedure:
 *	SIGPIPE handler - SIGPIPE means fvwm is dying
 *
 ***********************************************************************/
void DeadPipe(int nonsense)
{
  exit(0);
}

static char *MkDef(char *name, char *def)
{
  char *cp = NULL;
  int n;

  /* Get space to hold everything, if needed */

  n = EXTRA + strlen(name) + strlen(def);
  cp = safemalloc(n);

  sprintf(cp, "#define %s %s\n",name,def);

  return(cp);
}

static char *MkNum(char *name,int def)
{
  char num[20];

  sprintf(num, "%d", def);

  return(MkDef(name, num));
}
